### PR TITLE
[WIP][EMB-180] Prerender service

### PR DESCRIPTION
**Pull request by aaxelb**
## Purpose
Set `window.prerenderReady` after all loading on the page is finished.

## Summary of Changes
Add a `prerender` service, with `waitOn` and `finished` methods. If a route/controller/component loads initial data, it should call `waitOn('some key')` before, and `finished('some key')` after. `prerender` will set `window.prerenderReady` when everything that has called `waitOn` has also called `finished`.

Added proof-of-concept uses for route, controller, and component.

## Side Effects / Testing Notes
It might be annoying to add this to everything that loads data. It's only 2-3 lines, but it's yet another things to remember.

## Ticket

https://openscience.atlassian.net/browse/EMB-180

# Reviewer Checklist

- [ ] meets requirements
- [ ] easy to understand
- [ ] DRY
- [ ] testable and includes test(s)
- [ ] changes described in `CHANGELOG.md`

### CODE REVIEW

File: app/prerender/service.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import { A } from '@ember/array';
+import { assert } from '@ember/debug';
```
Comment (jamescdavis) [inline app/prerender/service.ts]: Yay for ember-debug!

File: app/prerender/service.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        assert('Already waiting on that key', !pending.includes(key));
+        pending.pushObject(key);
+    }
+
+    finished(this: Prerender, key: string): void {
+        if (window.prerenderReady) {
+            return;
+        }
+        const pending = this.get('pending');
+        assert('Call `waitOn` before `finished`', pending.includes(key));
```
Comment (jamescdavis) [inline app/prerender/service.ts]: Did you mean `!pending.includes(key)`?

Comment (aaxelb) [reply app/prerender/service.ts]: I'm thinking that when you call `finished`, the key should already be in `pending` from when you called `waitOn` earlier.

Comment (jamescdavis) [reply app/prerender/service.ts]: Ignore me, I wasn't thinking clearly about how `assert()` works.

**Comment by aaxelb** [comment]
Closing in favor of #140

**Pull request closed without merging**